src/drivers/ata-pci.c

```diff
@@ -66,36 +134,209 @@ static const char *ata_pcivendor2str(pci_dev d)
 }
 #endif // ATA_DEBUG
 
-static struct ata *ata_pci_alloc(heap general, pci_dev d)
+static ata_pci ata_pci_alloc(heap general, heap contiguous, pci_dev d)
 {
-    struct ata *dev = ata_alloc(general);
+    ata_pci apci = allocate(general, sizeof(*apci));
+    assert(apci != INVALID_ADDRESS);
+    apci->ata = ata_alloc(general);
+    apci->prdt = allocate(contiguous, PRDT_SIZE);
+    assert(apci->prdt != INVALID_ADDRESS);
     ata_debug("%s: %s ATA controller\n", __func__, ata_pcivendor2str(d));
-    return dev;
+    return apci;
+}
+
+void ata_pci_dealloc(heap general, heap contiguous, ata_pci apci)
+{
+    ata_debug("%s\n", __func__);
+    deallocate(contiguous, apci->prdt, PRDT_SIZE);
+    ata_dealloc(apci->ata);
+    deallocate(general, apci, sizeof(*apci));
 }
 
-closure_function(2, 1, boolean, ata_pci_probe,
-                 heap, general, storage_attach, a,
+static boolean ata_pci_service_req(ata_pci apci, ata_pci_req req)
+{
+    ata_debug("%s: %R %v\n", __func__, req->remain, req->s);
+    if (!is_ok(req->s))
+        goto done;
+    u64 byte_count = range_span(req->remain) * SECTOR_SIZE;
+    if (byte_count == 0)
+        goto done;
+    u64 buf_phys;
+    int prd_count = 0;
+    while (byte_count > 0 && prd_count < PRDT_ENTRIES) {
+        buf_phys = physical_from_virtual(req->buf);
+        if (buf_phys >= U64_FROM_BIT(32))
+            break;
+
+        /* Ensure the buffer doesn't cross a 64K boundary. */
+        u64 boundary = pad(buf_phys, 64 * KB);
+        if (boundary == buf_phys)
+            boundary += 64 * KB;
+        u64 len = MIN(byte_count, boundary - buf_phys);
+
+        ata_debug("%s: PRD address 0x%08x, len %ld\n", __func__, buf_phys, len);
+        req->buf += len;
+        byte_count -= len;
+        if (len == U64_FROM_BIT(16))
+            len = 0;
+        apci->prdt[prd_count].buf_addr = buf_phys;
+        apci->prdt[prd_count].byte_count = len;
+        apci->prdt[prd_count].ctrl = 0;
+        prd_count++;
+    }
+    if (prd_count > 0) {
+        apci->prdt[prd_count - 1].ctrl = PRD_LAST_ENTRY;
+        range blocks = irange(req->remain.start,
+            req->remain.end - byte_count / SECTOR_SIZE);
+        ata_debug("%s: starting DMA for %R\n", __func__, blocks);
+        pci_bar_write_4(&apci->bmr, ATA_BMR_PRDT(ATA_PRIMARY), apci->prdt_phys);
+        if (!ata_io_cmd_dma(apci->ata, req->write, blocks)) {
+            req->s = timm("result", "ATA command failed");
+            goto done;
+        }
+        req->remain = irange(blocks.end, req->remain.end);
+        write_barrier();
+        u8 cmd = ATA_BMR_CMD_START | (!req->write ? ATA_BMR_CMD_READ : 0);
+        pci_bar_write_1(&apci->bmr, ATA_BMR_CMD(ATA_PRIMARY), cmd);
+        return false;
+    }
+
+    /* Couldn't use DMA, fall back to PIO. */
```

Comment:
If the request stays in the reqs list, it won't be serviced (the next call to ata_pci_service_req() won't happen) until the next interrupt arrives (because when ata_pci_service_req() returns false, ata_pci_service_reqs() stops calling it), at which point the PIO transfer would occur. If we were to issue the pio request within the loop, we would have to mess with the status handler to check whether the I/O was successful and only continue the loop if there are no errors; this seems to me more complicated than the current approach.